# Correção exercícios Aula 13

### Leitura dos dados

In [27]:
# Importar a biblioteca necessária para leitura e manipulação de tabelas com o apelido pd
import pandas as pd

# Criação da variável df e leitura da base de dados
df = pd.read_csv('preco_carros.csv')

c:\Users\apcs1\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (7,12) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


### Análise da base de Dados

In [28]:
# Exibe as 3 linhas iniciais da tabela
df.head(3)

maker    model   mileage  manufacture_year  engine_displacement  \
0   ford   galaxy  151000.0            2011.0               2000.0   
1  skoda  octavia  143476.0            2012.0               2000.0   
2    bmw      NaN   97676.0            2010.0               1995.0   

   engine_power body_type color_slug  stk_year transmission  door_count  \
0         103.0       NaN        NaN       NaN          man         5.0   
1          81.0       NaN        NaN       NaN          man         5.0   
2          85.0       NaN        NaN       NaN          man         5.0   

   seat_count fuel_type  price_eur  
0         7.0    diesel   10584.75  
1         5.0    diesel    8882.31  
2         5.0    diesel   12065.06

In [29]:
# Descrição das colunas numéricas
df.describe().T

count          mean           std      min       25%  \
mileage              3190328.0  1.158140e+05  3.422508e+05     0.00  18800.00   
manufacture_year     3182334.0  2.000871e+03  8.172588e+01     0.00   2004.00   
engine_displacement  2809498.0  2.043958e+03  1.973958e+03     0.00   1400.00   
engine_power         2998035.0  9.846796e+01  4.907309e+01     1.00     68.00   
stk_year              536105.0  2.021473e+03  1.404767e+02  2015.00   2017.00   
door_count           2462846.0  4.052180e+00  9.911850e-01     0.00      4.00   
seat_count           2265813.0  4.869694e+00  1.212611e+00     0.00      5.00   
price_eur            3552912.0  1.625812e+06  2.025622e+09     0.04   1295.34   

                          50%        75%           max  
mileage              86415.00  158025.00  9.999999e+06  
manufacture_year      2009.00    2013.00  2.017000e+03  
engine_displacement   1798.00    1997.00  3.276700e+04  
engine_power            86.00     110.00  2.237000e+03  
stk_year              2017.00    2018.00  9.990000e+03  
door_count               4.00       5.00  7.700000e+01  
seat_count               5.00       5.00  5.170000e+02  
price_eur             7364.91   16284.23  2.706149e+12

In [30]:
# Descrição das colunas Textuais
df.describe(include = 'O').T

count unique       top     freq
maker         3033997     46     skoda   313830
model         2419551   1012   octavia   129563
body_type     2429998      9     other  1964289
color_slug     209501     14     black    42931
transmission  2811282      2       man  2021990
fuel_type     1705306      5  gasoline   902222

In [31]:
# Tipo de dados das colunas
df.dtypes

maker                   object
model                   object
mileage                float64
manufacture_year       float64
engine_displacement    float64
engine_power           float64
body_type               object
color_slug              object
stk_year               float64
transmission            object
door_count             float64
seat_count             float64
fuel_type               object
price_eur              float64
dtype: object

In [32]:
# Quantidade de Valalores nulos por colunas
df.isnull().sum()

maker                   518915
model                  1133361
mileage                 362584
manufacture_year        370578
engine_displacement     743414
engine_power            554877
body_type              1122914
color_slug             3343411
stk_year               3016807
transmission            741630
door_count             1090066
seat_count             1287099
fuel_type              1847606
price_eur                    0
dtype: int64

In [33]:
# Porcentagem de Valalores nulos por colunas
round(df.isnull().sum() * 100 / len(df))

maker                  15.0
model                  32.0
mileage                10.0
manufacture_year       10.0
engine_displacement    21.0
engine_power           16.0
body_type              32.0
color_slug             94.0
stk_year               85.0
transmission           21.0
door_count             31.0
seat_count             36.0
fuel_type              52.0
price_eur               0.0
dtype: float64

# Tratamento de Valores Nulos
Todas as colunas de atributos possuem valores nulos.

Para tratar valores nulos temos que verificar 2 pontos importantes:
* Porcentagem de valores nulos;
* Tipo dos dados. 

### Porcentagem de valores nulos:
* Com porcentagens acima de 50% é aceitável que se exclua a coluna, mas com valores acima de 80% é super recomendável que se exclua;
* Com porcentagens entre 1% e 80% podemos substituir os valores nulos pela (Média, Mediana ou Moda - Verficar o tipo de dados;
* Com valores abaixo de 1% podemos excluir as linhas com valores nulos ou substituir os valores nulos pela (Média, Mediana ou Moda - Verficar o tipo de dados).

### Tipo de Dados
* Dados numéricos (int ou float) podemos substituir por:
    * Média - Soma de todos os valores dividido pela quantidade de valores;
    * Mediana - Todos Valores "listados" em ordem crescente, a mediana é o valor que aparece na metade da lista;
    * Moda - O valor que mais aparece

* Dados Textuais, podemos substituir pela moda

# 1 Excluir colunas com porcentagens altas de valores nulos
color_slug             94%

stk_year               85%

fuel_type              52%


In [34]:
# Apaga as colunas 'color_slug', 'stk_year' e 'fuel_type'
df.drop(['color_slug', 'stk_year', 'fuel_type'], axis=1, inplace = True)

# 2 - Tratar valores nulos das colunas numéricas

Quilometragem - Interessante substituir pela média (neste caso não vejo tanta necessidade de remover as casas decimais mas irei remover)
* mileage ------------- float64

Ano de Fabricação - Interessante substituir pela moda (Sempre é interessante verificar com o `valeu_counts()` quantas vezes a moda aparece e quantos valores únicos existem na coluna)
* manufacture_year ----- float64

Cilindradas do Motor - Usar moda (neste caso não teve muito critério...)
* engine_displacement -- float64

Quantos cavalos o carro possui - Usar média (neste caso não teve muito critério...), importante lembrar de remover as casas decimais, pois neste caso, não existe carro com, por exemplo, 100.7 cavalos...
* engine_power --------- float64

Quantidade de Portas - Usar moda (Sempre é interessante verificar com o `valeu_counts()` quantas vezes a moda aparece e quantos valores únicos existem na coluna)
* door_count ----------- float64

Quantidade de assentos - Tanto a moda, quanto mediana e média arredondada retornam o mesmo valor
* seat_count ----------- float64


In [35]:
# Substitui valores nulos de kilometragem pela média arredondada
df.loc[df['mileage'].isnull(), 'mileage'] = round(df.mileage.median())

# Substitui valores nulos de ano de fabricação pela moda
df.loc[df['manufacture_year'].isnull(), 'manufacture_year'] = df.manufacture_year.mode()[0]

# Substitui valores nulos de cilindradas pela moda
df.loc[df['engine_displacement'].isnull(), 'engine_displacement'] = df.engine_displacement.mode()[0]

# Substitui valores nulos de quantidade de cavalos pela média arredondada
df.loc[df['engine_power'].isnull(), 'engine_power'] = round(df.engine_power.median())

# Substitui valores nulos de quantidade de portas pela moda
df.loc[df['door_count'].isnull(), 'door_count'] = df.door_count.mode()[0]

# Substitui valores nulos de assentos pela moda
df.loc[df['seat_count'].isnull(), 'seat_count'] = df.seat_count.mode()[0]

# 3 - Tratar valores nulos das colunas textuais
maker ------- object

model ------- object

body_type ------- object

transmission ------- object

No caso das colunas textuais, não conswguimos usar a média nem a mediana (pois são cálculos matemáticos que só podem ser aplicados a valores numéricos), então iremos substituir os valores nulos pela moda (valor que mais aparece)

In [36]:
# Substitui valores nulos de fabricante pela moda
df.loc[df['maker'].isnull(), 'maker'] = df.maker.mode()[0]

# Substitui valores nulos de modelo pela moda
df.loc[df['model'].isnull(), 'model'] = df.model.mode()[0]

# Substitui valores nulos de tipo de carro pela moda
df.loc[df['body_type'].isnull(), 'body_type'] = df.body_type.mode()[0]

# Substitui valores nulos de trnsmissão pela moda
df.loc[df['transmission'].isnull(), 'transmission'] = df.transmission.mode()[0]

In [37]:
df.body_type.value_counts() * 100 / len(df)

other           86.892189
compact          6.809851
coupe            1.986427
stationwagon     1.967260
van              0.880967
offroad          0.634662
sedan            0.553602
convertible      0.150074
transporter      0.124968
Name: body_type, dtype: float64

In [38]:
# Quantidade de Valalores nulos por colunas
df.isnull().sum()

maker                  0
model                  0
mileage                0
manufacture_year       0
engine_displacement    0
engine_power           0
body_type              0
transmission           0
door_count             0
seat_count             0
price_eur              0
dtype: int64

# 4 - Converter dados textuais para numéricos.

maker ------- object

model ------- object

body_type ------- object

transmission ------- object

Para facilitar, iremos usar a biblioteca label encoder do sklearn.

In [39]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

df['maker'] = le.fit_transform(df['maker'])
df['model'] = le.fit_transform(df['model'])
df['body_type'] = le.fit_transform(df['body_type'])
df['transmission'] = le.fit_transform(df['transmission'])

In [40]:
# Verifica o tipo de dados das colunas (recebemos apenas tipos numéricos)
df.dtypes

maker                    int32
model                    int32
mileage                float64
manufacture_year       float64
engine_displacement    float64
engine_power           float64
body_type                int32
transmission             int32
door_count             float64
seat_count             float64
price_eur              float64
dtype: object

# 5 - Classe Alvo (o que queremos prever)

Geralmente não existem muistos valores únicos na nossa classe alvo. Como vimos em toso os outros exemplos até aqui, todas as bases tinham apenas 2 valores, sim e não, 1 e 0...

Não existe problema e existir mais de 2 valores únicos, entretanto, muitos valores únicos pode acabar gerando uma acurácia muito baixa, visto que o modelo terá que prever valores em uma faixa muito grande de valores.

Geralmente estes casos são quando a classe alvo é idade, valor fnanceiro... Nestes casos, é recomendado que sejam criadas faixas de valores, como por exemplo, substituir os valores de idade (de 0 a 100) por faixas de idade como:

1 = de 0 a 10 anos

2 = de 11 a 20 anos

3 = de 21 a 30 anos

4 = de 31 a 40 anos

5 = de 41 a 50 anos

6 = de 51 a 60 anos

7 = de 61 a 70 anos

8 = de 71 a 80 anos

9 = de 81 a 90 anos

10 = acima de 90 anos

Desta forma, conseguimos reduzir a 1 décimo a quatidade de valores únicos na classe alvo (ao invés de valores de 0 a 100, valores de 1 a 10).

Nesta base e dados, temos realizar este tipo de coreção também (são 227305 valores únicos)

# 6 - Tratamento de valores errados
As vezes a base de dados pode ter valores errados, por erro de digitação humana ou algum outro motivo.

Outro ponto imporane, é verificar quais os valros máximos e mínimos da coluna.


In [41]:
# Quantidade de valores únicos na classe alvo
df.price_eur.nunique()

227305

In [42]:
# Valor máximo da coluna price_eur
max(df.price_eur)

2706149053064.4

In [43]:
# Valor mínimo da coluna price_eur
min(df.price_eur)

0.04

Repare que o na teoria, o carro mais barato é de 0.04 centavos e o mais caro 2706149053064.4 euros...

Certamente existem erros nestes valores. Assim como valores nulos, temos 2 opções aqui:
* Apagar os registros (linhas onde os valores estão errado)
* Substituir os valores errados pela média, mediana ou moda da coluna.

Eu particularmente prefiro excluir as linhas onde existem erros na classe alvo.

Para isto, vamos definir um valor máximo que um carro pode custar e um valor mínimo que o carro pode custar.

Valor máximo = 400.000 euros
Valor mínimo= 400 euros

Com estes valores definidos, podemos usar o loc para selecionar somente os carros dentro desta faixa de valores

In [44]:
# Mostra o total de carros na base de dados
df.shape[0]

3552912

In [45]:
# Mostra o total de carros na base de dados com nosso filtro aplicado
df.loc[(df['price_eur']>=400) & (df['price_eur']<=400000)].shape[0]

3496821

In [46]:
# Mostra o total de carros na base de dados que estão fora do nosso filtro - quantidade de carros que serão removidos
df.shape[0] - df.loc[(df['price_eur']>=400) & (df['price_eur']<=400000)].shape[0]

56091

In [47]:
# Atribuímos à variável df o df com filtro aplicado
df = df.loc[(df['price_eur']>=400) & (df['price_eur']<=400000)]

Ainda assim, existem muitos valores únicos nesta classe, vamos usar um loop for com a função loc dentro para solucionar este problema

In [48]:
#cria um contador
cont = 1

# Criamos um loop que começa em 2000, vai até 400001 e and e 2000 em 2000
for var in range(4000, 400001, 4000):
    # Faz um filtro que filtra os valores de 4000 em 4000 e altera o valor para o valor do contador
    df.loc[(df['price_eur']<=var)& (df['price_eur']>var-4000), 'price_eur'] = cont
    cont +=1

desta forma, os valores foram substituidos por:

1 = valores entre 400 e 4000

2 = valores entre 4000,01 e 8000

3 = valores entre 8000,01 e 12000

4 = valores entre 12000,01 e 16000

5 = valores entre 16000,01 e 20000



.

.

.

Ou seja, em cálculo matemáticos, os valores estão:

(novo valor * 4000) -4000 a (novo valor * 4000)

(2 * 4000) -4000 a (2 * 4000)

4000 a 8000

------------------------------------------------

(novo valor * 4000) -4000 a (novo valor * 4000)

(5 * 4000) -4000 a (5 * 4000)

16000 a 20000

In [49]:
# Quantidade de valores únicos na classe alvo
df.price_eur.nunique()

100

In [50]:
df

maker  model   mileage  manufacture_year  engine_displacement  \
0           11    484  151000.0            2011.0               2000.0   
1           38    679  143476.0            2012.0               2000.0   
2            4    679   97676.0            2010.0               1995.0   
3           38    450  111970.0            2004.0               1200.0   
4           38    450  128886.0            2004.0               1200.0   
...        ...    ...       ...               ...                  ...   
3552907     38    774   54000.0            2013.0               1200.0   
3552908     38    452   86415.0            2000.0               1968.0   
3552909     38    679  230000.0            2006.0               1900.0   
3552910     38    450   86415.0            2001.0               1968.0   
3552911     27    679   86415.0            2015.0               1968.0   

         engine_power  body_type  transmission  door_count  seat_count  \
0               103.0          4             1         5.0         7.0   
1                81.0          4             1         5.0         5.0   
2                85.0          4             1         5.0         5.0   
3                47.0          4             1         5.0         5.0   
4                47.0          4             1         5.0         5.0   
...               ...        ...           ...         ...         ...   
3552907          63.0          4             1         4.0         5.0   
3552908          50.0          4             1         4.0         5.0   
3552909         100.0          4             1         4.0         5.0   
3552910          86.0          4             1         4.0         5.0   
3552911          86.0          4             1         4.0         5.0   

         price_eur  
0              3.0  
1              3.0  
2              4.0  
3              1.0  
4              1.0  
...            ...  
3552907        1.0  
3552908        1.0  
3552909        1.0  
3552910        1.0  
3552911        1.0  

[3496821 rows x 11 columns]

In [51]:
# Valor máximo da coluna price_eur
max(df.price_eur)

100.0

In [52]:
# Valor máximo da coluna price_eur
min(df.price_eur)

1.0

Desta forma, temos valores de 0 a 100 (Ainda é uma quantidade relativamente alta de valores únicos para a classe alvo), mas já é bem menos do que havia.

# 7 - Normalizar os dados
A normalização mais utilizada conhecida é a min max, onde o maior valor se tornará 1 e o menor 0 e a sua fórmula é dada por:

``x' = (x - min(x)) / (max(x) - min(x))``

Um exemplo prático pode ser a normalização da seguinte lista:

[-8, 5, 12, -1, 15]

Para cada valor da lista iremos aplpicar a fórmula, com os menores (min) e maiores (max) valores da lista

min = -8

max = 15

(-8 - (-8)) / (15 - (-8)) = 0

(5 - (-8)) / (15 - (-8)) = 0.5652173913043478

(12 - (-8)) / (15 - (-8)) = 0.8695652173913043

(-1 - (-8)) / (15 - (-8)) = 0.30434782608695654

(15 - (-8)) / (15 - (-8)) = 1

nova lista = [0, 0.5652173913043478, 0.8695652173913043, 0.30434782608695654, 1]

In [ ]:
# Aplicação da mesma fórmula para a lista, na prática
[(x - min([-8, 5, 12, -1, 15])) / (max([-8, 5, 12, -1, 15]) - min([-8, 5, 12, -1, 15])) for x in [-8, 5, 12, -1, 15]]

In [53]:
# Normaliza os dados no df_norm
df_norm = (df - df.min()) / (df.max() - df.min())

Após normalizar, geralmente removemos a normalização da nossa classe alvo, entretanto, ao invés de fazermos isto, também podemos definir o nosso y referente ao nosso dataframe original sem normalização.

OBS: Devido a conceitos matemáticos, por padrão o X é maiúsculo e o y minúsculo, mas note, isto são apenas variáveis, então na prática, podemos usar o nome de variáveis que desejarmos...

In [54]:
# X (X Maiúsculo) recebe todas as colunas normalizadas até a penúltima (visto que a última é nossa classe alvo)
X = df_norm[list(df_norm.columns[0:-1])]

# y (y minúsculo) recebe a nossa classe alvo da tabela que não foi normalizada
y = df['price_eur']

# 6 - Separar os dados em treinamento e teste

Para isto, iremos usar, também do sklearn, a biblioteca "train_test_split"

In [55]:
# Biblioteca de calcar acurácia
from sklearn.metrics import accuracy_score

from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.40)

In [56]:
from sklearn import svm
# Biblioteca de calcar acurácia
from sklearn.metrics import accuracy_score

clf = svm.SVC()
clf.fit(train_X, train_y)

pred_y = clf.predict(test_X)

accuracy_score(test_y, pred_y)

Após longos tempos treinando modelos, é interessante salvar o modelo treinado para aplicar posteriormente.

In [ ]:
# Salvar o modelo
# Se preciso, execute !pip install pickle
import pickle

arquivo = 'modelo_treinado.sav'
pickle.dump(clf, open(arquivo, 'wb'))
 
# Algum tempo depois...
 
# Carrega o modelo
modelo = pickle.load(open(arquivo, 'rb'))

pred_y = clf.predict(test_X)

accuracy_score(test_y, pred_y)